src/ruis/render/scene/gltf_loader.hxx

```diff
@@ -0,0 +1,219 @@
+	std::vector<utki::shared_ref<node>> nodes; // all nodes that are listed in gltf file
+	std::vector<utki::shared_ref<mesh>> meshes; // meshes.
+	std::vector<utki::shared_ref<accessor>> accessors; // accessors.
+	std::vector<utki::shared_ref<buffer_view>> buffer_views; // bv's
+
+	std::vector<utki::shared_ref<material>> materials;
+	std::vector<utki::shared_ref<ruis::render::texture_2d>> textures;
+	std::vector<utki::shared_ref<sampler_l>> samplers;
+	std::vector<utki::shared_ref<image_l>> images;
+
+	std::vector<std::vector<uint32_t>> child_indices; // storage for node child hierarchy (only during loading stage)
+
+	template <typename tp_type>
+	void create_vertex_buffer_float(
+		utki::shared_ref<ruis::render::accessor>,
+		utki::span<const uint8_t> buffer,
+		uint32_t acc_count,
+		uint32_t acc_stride
+	);
+
+	template <typename tp_type>
+	std::vector<utki::shared_ref<tp_type>> read_root_array(
+		std::function<tp_type(const jsondom::value& j)> read_func,
+		const jsondom::value& root_json,
+		const std::string& name
+	);
+
+	template <typename tp_type>
+	utki::shared_ref<ruis::render::vertex_array> create_vao_with_tangent_space(
+		utki::shared_ref<accessor> index_accessor,
+		utki::shared_ref<accessor> position_accessor,
+		utki::shared_ref<accessor> texcoord_0_accessor,
+		utki::shared_ref<accessor> normal_accessor
+	);
+
+	utki::shared_ref<buffer_view> read_buffer_view(const jsondom::value& buffer_view_json);
+	utki::shared_ref<accessor> read_accessor(const jsondom::value& accessor_json);
+	utki::shared_ref<mesh> read_mesh(const jsondom::value& mesh_json);
+	utki::shared_ref<node> read_node(const jsondom::value& node_json);
+	utki::shared_ref<scene> read_scene(const jsondom::value& scene_json);
+
+	utki::shared_ref<image_l> read_image(const jsondom::value& image_json);
+	utki::shared_ref<sampler_l> read_sampler(const jsondom::value& sampler_json);
+	utki::shared_ref<ruis::render::texture_2d> read_texture(const jsondom::value& texture_json);
+	utki::shared_ref<material> read_material(const jsondom::value& material_json);
+
+public:
+	utki::shared_ref<scene> load(const papki::file& fi);
+	gltf_loader(ruis::render::factory& factory_v);
+};
+
+struct buffer_view // currently we support only one data buffer, the single data buffer located in the .glb file
+{
+	// utki::span<uint8_t> buffer; // for further development use
+	uint32_t byte_length;
+	uint32_t byte_offset;
+	uint32_t byte_stride;
+
+	enum class target {
+		undefined = 0,
+		array_buffer = 34962,
+		element_array_buffer = 34963
+	} target_v;
+
+	buffer_view(uint32_t byte_length, uint32_t byte_offset, uint32_t byte_stride, target target_v) :
+		byte_length(byte_length),
+		byte_offset(byte_offset),
+		byte_stride(byte_stride),
+		target_v(target_v)
+	{}
+};
+
+struct accessor {
+	utki::shared_ref<buffer_view> bv;
+	uint32_t count;
+	uint32_t byte_offset;
+
+	enum class type {
+		undefined = 0,
+		scalar = 1,
+		vec2 = 2,
+		vec3 = 3,
+		vec4 = 4,
+		mat2 = 5, // whoa
```

Comment:
add more description about what is so special about this enum item. The "whoa" doesn't tell me anything.

Reply:
The idea was to use this type for both unique ID and data size, but we don't support matrices as input data anyway.
Changed comment to reflect this

Reply:
actually, removed the comment